src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
```

Comment:
Is there some way to avoid having this as a public class? Adding some static methods to torch.?? Not sure, but it feels like it should be possible to avoid it being public, and to only have DisposeScope exposed.

Reply:
The only reason is to have a thread safe way of finding out how many tensors have been created on the thread. But that could definitely be moved somewhere else, the tensor class could keep track of it with some thread static magic. 

        [Fact]
        public void DisposeScopesCanBeNestled()
        {
            torch.Tensor data = torch.rand(10, 10);
            var preTotalCount = DisposeScopeManager.Singleton.TotalCount;

            using (torch.NewDisposeScope()) {
                var t1 = data * data + data;

                var innerCount = DisposeScopeManager.Singleton.TotalCount;
                using (torch.NewDisposeScope()) {
                    var t2 = data + data - t1;
                }

                // Inner scope was disposed
                Assert.Equal(DisposeScopeManager.Singleton.TotalCount, innerCount);
            }

            // It was all disposed
            Assert.Equal(DisposeScopeManager.Singleton.TotalCount, preTotalCount);
        }

Reply:
I've made the singleton internal and moved the count to DisposeScopeManager.ThreadTotalLiveCount